### PR TITLE
18zoo add tickets

### PR DESCRIPTION
**Pull request by carlorusso**
Addition to 18ZOO:
- add ZOOTickets on game creation
- companies can be bought from players
- companies (only OR or ZOOTicket from owner player) can be bought by corporation
- ZOOTicket can be sold by corporation (price to player, difference to corporation)
- ZOOTickets value changes every round
- Player orders is higher cash + current order (but not on Draft)

### CODE REVIEW

File: lib/engine/game/g_18_zoo.rb [hunk 1]
```diff
@@ -546,6 +546,42 @@ def self.title
         '18ZOO'
       end
 
+      def purchasable_companies(entity = nil)
+        entity ||= @round.current_operator
+        return [] unless entity && (entity.corporation? || entity.player?)
```
Comment (tobymao) [inline lib/engine/game/g_18_zoo.rb:551]: don't use unless with multiple conditoions change it to an if

File: lib/engine/step/buy_company.rb [hunk 2]
```diff
@@ -104,6 +99,16 @@ def round_state
       def setup
         @blocks = @opts[:blocks] || false
       end
+
+      def pay(entity, owner, price, company)
+        entity.spend(price, owner.nil? ? @game.bank : owner)
```
Comment (tobymao) [inline lib/engine/step/buy_company.rb:104]: owner || @game.bank

File: lib/engine/step/g_18_zoo/buy_company.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            @log << "#{corporation.name} earns #{@game.format_currency(company.value - price)}"
+
+            company.close!
+          else
+            super
+          end
+        end
+
+        def pay(entity, owner, price, company)
+          entity.spend(price, owner.nil? ? @game.bank : owner) if price.positive?
```
Comment (tobymao) [inline lib/engine/step/g_18_zoo/buy_company.rb:39]: owner || banke

File: lib/engine/step/g_18_zoo/buy_sell_par_shares.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+require_relative '../buy_sell_par_shares'
+
+module Engine
+  module Step
+    module G18ZOO
+      class BuySellParShares < BuySellParShares
+        def actions(entity)
```
Comment (tobymao) [inline lib/engine/step/g_18_zoo/buy_sell_par_shares.rb:9]: why did you need to implement this method?

Comment (carlorusso) [reply lib/engine/step/g_18_zoo/buy_sell_par_shares.rb:9]: the original contains
`actions << 'buy_company' unless purchasable_companies(entity).empty?`
but the `purchasable_companies` is used for both showing the buyable companies on the top and on the bottom
Screenshot 2021-01-31 at 21 49 06
I don't want to add `can_buy_companies_from_other_players` but without that I'll never be able to show the companies.

Maybe we could change the behaviour to be able to buy from bank without `can_buy_companies_from_other_players` ?

File: lib/engine/game/g_18_zoo.rb (hunk 1 above)
Comment (carlorusso) [reply lib/engine/game/g_18_zoo.rb]: done in 955e5710

File: lib/engine/step/buy_company.rb (hunk 2 above)
Comment (carlorusso) [reply lib/engine/step/buy_company.rb]: done in 955e5710

File: lib/engine/step/g_18_zoo/buy_company.rb (hunk 3 above)
Comment (carlorusso) [reply lib/engine/step/g_18_zoo/buy_company.rb]: done in 955e5710

**Review by tobymao — APPROVED** [review]
